sdks/java/core/src/test/java/org/apache/beam/sdk/util/HistogramDataTest.java

```diff
@@ -200,4 +202,134 @@ public void testIncrementBucketCountByIndex() {
+    // Buckets will be:
+    // Index        Range
+    // Underflow    (-inf, 0)
+    // 0            [0, 2)
+    // 1            [2, 2^2]
+    // i            [2^i, 2^(i+1))
+    HistogramData data = HistogramData.exponential(0, 20);
+
+    data.record(-1);
+    assertThat(data.getBottomBucketCount(), equalTo(1L));
+
+    data.record(0, 1);
+    assertThat(data.getCount(0), equalTo(2L));
+
+    data.record(4, 5, 6, 7);
+    assertThat(data.getCount(2), equalTo(4L));
+
+    for (int i = 32; i < 64; i++) {
+      data.record(i);
+    }
+    assertThat(data.getCount(5), equalTo(32L));
+
+    for (int i = IntMath.pow(2, 16); i < IntMath.pow(2, 16) + 100; i++) {
+      data.record(i);
+    }
+    assertThat(data.getCount(16), equalTo(100L));
+
+    Long expectedTotalCount = Long.valueOf(100 + 32 + 4 + 2 + 1);
+    assertThat(data.getTotalCount(), equalTo(expectedTotalCount));
+  }
+
+  @Test
+  public void testNegativeScaleBucket() {
+    // Buckets will be:
+    // Index        Range
+    // Underflow    (-inf, 0)
+    // 0            [0, 4)
+    // 1            [4, 4^2]
+    // i            [4^i, 4^(i+1))
+    HistogramData data = HistogramData.exponential(-1, 20);
+
+    data.record(-1);
+    assertThat(data.getBottomBucketCount(), equalTo(1L));
+
+    data.record(0, 1, 2);
+    assertThat(data.getCount(0), equalTo(3L));
+
+    data.record(16, 17, 32, 33, 62, 63);
+    assertThat(data.getCount(2), equalTo(6L));
+
+    for (int i = IntMath.pow(4, 5); i < IntMath.pow(4, 5) + 20; i++) {
+      data.record(i);
+    }
+    assertThat(data.getCount(5), equalTo(20L));
+
+    Long expectedTotalCount = Long.valueOf(20 + 6 + 3 + 1);
+    assertThat(data.getTotalCount(), equalTo(expectedTotalCount));
+  }
+
+  @Test
+  public void testBucketSize() {
+    HistogramData data = HistogramData.exponential(0, 20);
+    assertThat(data.getBucketType().getBucketSize(0), equalTo(2.0));
+    // 10th bucket contains [2^10, 2^11).
+    assertThat(data.getBucketType().getBucketSize(10), equalTo(1024.0));
+
+    data = HistogramData.exponential(1, 20);
+    assertThat(data.getBucketType().getBucketSize(0), equalTo(Math.sqrt(2)));
+    // 10th bucket contains [2^5, 2^5.5).
+    assertThat(data.getBucketType().getBucketSize(10), closeTo(13.2, .1));
+
+    data = HistogramData.exponential(-1, 20);
+    assertThat(data.getBucketType().getBucketSize(0), equalTo(4.0));
+    // 10th bucket contains [2^20, 2^22).
+    assertThat(data.getBucketType().getBucketSize(10), equalTo(3145728.0));
+  }
+
+  @Test
+  public void testNumBuckets() {
+    // Validate that numBuckets clipping WAI.
+    HistogramData data = HistogramData.exponential(0, 200);
+    assertThat(data.getBucketType().getNumBuckets(), equalTo(32));
+
+    data = HistogramData.exponential(3, 500);
+    assertThat(data.getBucketType().getNumBuckets(), equalTo(32 * 8));
+
+    data = HistogramData.exponential(-3, 500);
+    assertThat(data.getBucketType().getNumBuckets(), equalTo(4));
+  }
```

Comment:
since `HistogramData.exponential(...)` creates a new instance of `HistogramData`, how about just declaring a new variable for every instance for clarity?

It is also ok to have even smaller tests with narrower assertions if each instance is testing a different part of the behavior of the method (in this case`getNumBuckets`) go/tott/648

Reply:
Done.